experiment/doppler/main.go

```diff
@@ -0,0 +1,264 @@
+		for _, job := range jobs {
+			// populate jobMap
+			if _, ok := jobMap[job.Job]; !ok {
+				jobMap[job.Job] = make(map[string][]string)
+			}
+			if _, ok := jobMap[job.Job][job.PullSHA]; !ok {
+				jobMap[job.Job][job.PullSHA] = make([]string, 0)
+			}
+			jobMap[job.Job][job.PullSHA] = append(jobMap[job.Job][job.PullSHA], job.State)
+			// populate commitMap
+			if _, ok := commitMap[job.PullSHA]; !ok {
+				commitMap[job.PullSHA] = make(map[string][]string)
+			}
+			if _, ok := commitMap[job.PullSHA][job.Job]; !ok {
+				commitMap[job.PullSHA][job.Job] = make([]string, 0)
+			}
+			commitMap[job.PullSHA][job.Job] = append(commitMap[job.PullSHA][job.Job], job.State)
+		}
+		jobCommits := make(map[string]int)
+		jobFlakes := make(map[string]int)
+		for job, commits := range jobMap {
+			jobCommits[job] = len(commits)
+			jobFlakes[job] = 0
+			for _, results := range commits {
+				hasFailure, hasSuccess := false, false
+				for _, state := range results {
+					if state == "success" {
+						hasSuccess = true
+					}
+					if state == "failure" {
+						hasFailure = true
+					}
+					if hasSuccess && hasFailure {
+						break
+					}
+				}
+				if hasSuccess && hasFailure {
+					jobFlakes[job]++
+				}
+			}
+		}
+
+		var flaky []jobResults
+		for job, flakeCount := range jobFlakes {
+			if jobCommits[job] < 10 {
+				continue
+			}
+			failChance := float64(flakeCount) / float64(jobCommits[job])
+			flaky = append(flaky, jobResults{
+				flakeCount:  flakeCount,
+				commitCount: jobCommits[job],
+				chance:      100 * failChance,
+				jobName:     job,
+			})
+		}
+
+		fmt.Printf("Certain flakes in %s:\n", repo)
+		sort.Sort(byChance(flaky))
+		for _, flake := range flaky {
+			fmt.Println(flake)
+			metrics.Set(repo, flake.jobName, flake.chance)
+		}
+
+		flaked := 0
+		for _, jobs := range commitMap {
+		out:
+			for _, results := range jobs {
+				hasFailure, hasSuccess := false, false
+				for _, state := range results {
+					if state == "success" {
+						hasSuccess = true
+					}
+					if state == "failure" {
+						hasFailure = true
+					}
+					if hasSuccess && hasFailure {
+						flaked++
+						break out
+					}
+				}
+			}
+		}
+		flakeChance := float64((flaked * 100.0)) / float64(len(commitMap))
+		metrics.Set(repo, total, flakeChance)
+		fmt.Printf("Commits that flaked in %s: %d/%d %.2f%%\n", repo, flaked, len(commitMap), flakeChance)
+	}
+	return nil
+}
+
+func main() {
+	log.SetFlags(log.LstdFlags | log.Lshortfile)
+	o := flagOptions()
+
+	// TODO: Refresh metrics
+	metrics, err := RegisterMetrics(o.prowURL, o.repo)
+	if err != nil {
+		log.Fatalf("could not load metrics: %v", err)
+	}
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+		for {
+			start := time.Now()
+			err := syncOne(o, metrics)
+			if err != nil {
+				log.Printf("error syncing: %v", err)
+				return
+			}
+			if o.runOnce {
+				return
+			}
+			log.Printf("Sync time: %v", time.Since(start))
```

Comment:
Observing races between this log and the fmt calls inside syncOne when running on GCE ..... :(